drv/lpc55-update-server/src/main.rs

```diff
@@ -573,6 +623,411 @@ impl ServerImpl<'_> {
+        // to check the image signature without crashing.
+        //
+        // Only an image that has a valid signature seen at rot-startup
+        // time will be copied by update-server to the stage0 partition.
+        //
+        // Note that both stage0 and stage0next images can be
+        // modified multiple times between resets.
+        // So, although we trust the bootstate(), we do not implicitly trust
+        // the current contents of stage0 or stage0next unless the image
+        // hash matches an image seen at rot-startup.
+        //
+        // The typical flow is that:
+        //   1. A new bootloader image is staged by update-server.
+        //   2. The RoT is reset.
+        //   3. Signatures are evaluated in rot-startup.
+        //   4. The validated staged image is copied by update-server
+        //      to the bootloader slot.
+        //   5. The RoT is rebooted into the new stage0 image.
+        //
+        // During step 4, there is a time where the RoT is not bootable.
+        // Failures during step 4 may require a service call or RMA to fix
+        // the system.
+        //
+        // Only one Gimlet, PSC, or SideCar in a rack should have their RoT
+        // bootloader updated at a time to minimize the blast-radius of
+        // failures such as a rack wide power issue.
+        //
+        // All of the stage0next slots in the rack can be updated (steps 1
+        // through 3) to ensure that RoT image signatures are valid before any
+        // system continues to step 4.
+        //
+        // TBD: While Failures up to step 3 do not adversly affect the RoT,
+        // resetting the RoT to evaluate signatures may be service affecting
+        // to the system depending on how the RoT and SP interact with respect
+        // to their reset handling and the RoT measurement of the SP.
+        // Appropriate measures should be taken to minimize customer impact.
+        //
+        // Note that after copying stage0next to stage0, but before step 5
+        // (reset), it is still possible to revert to the old stage0 by
+        // updating stage0next to the original stage0 contents that were
+        // validated reset and then copying those to stage0.
+        //
+        // It is assumed that a hash collision is not computaionally feasible
+        // for either the image hash done by rot-startup or used by the ROM
+        // signature routine.
+
+        // Read stage0next contents into RAM.
+        let staged = image_range(RotComponent::Stage0, SlotId::B);
+        let len = self.read_flash_image_to_cache(staged.0)?;
+        let bootloader = &self.fw_cache[..len / core::mem::size_of::<u32>()];
+
+        let mut hash = Sha3_256::new();
+        for page_start in (0..(bootloader.len() * 4)).step_by(512) {
+            // XXX
+            hash.update(
+                bootloader.as_bytes()[page_start..page_start + 512].as_ref(),
+            );
+        }
```

Comment:
style nit: i think we could loop over every 512-byte page by writing this, instead:
```suggestion
        for page in bootloader.as_bytes().chunks(512) {
            hash.update(page);
        }
```

which feels much clearer to me?

Reply:
Thanks.